python_modules/dagster/dagster/components/core/component_tree.py

```diff
@@ -377,14 +376,13 @@ def get_all_components(
         self,
         of_type: type[TComponent],
     ) -> list[TComponent]:
-        """Get all components from this context that are instance of the specified type."""
-        root_component = self.load_root_component()
-        if not isinstance(root_component, DefsFolderComponent):
-            raise Exception("Root component is not a DefsFolderComponent")
+        """Get all components from this context that are instance of the specified type.
+        Avoids loading components that are not of the specified type.
+        """
         return [
-            component
-            for component in root_component.iterate_components()
-            if isinstance(component, of_type)
+            cast("TComponent", self.load_component_at_path(path))
+            for path, decl in self._component_decl_tree().items()
+            if issubclass(decl.component_type, of_type)
```

Comment:
i think youll want to use safe_is_subclass

maybe promote this private copy to util
https://github.com/dagster-io/dagster/blob/master/python_modules/dagster/dagster/components/resolved/base.py#L392-L397